File: src/pipeline.ts
import { extractFromAtsJson, extractFromNotesTxt, extractFromRecruiterCsv } from './extractors';
import { mergeEvidence } from './merger';
import { defaultProjectionConfig, project } from './projector';
import { validateCustomProjection, validateDefaultProfile } from './validator';

export interface PipelineResult {
  canonical: Record<string, any>;
  projected: Record<string, any>;
  validation_errors: string[];
}

export function runPipeline(
  recruiterCsvPath?: string | null,
  atsJsonPath?: string | null,
  notesTxtPath?: string | null,
  config?: Record<string, any> | null,
  candidateId: string = 'candidate-001'
): PipelineResult {
  const evidence = [];
  if (recruiterCsvPath) evidence.push(...extractFromRecruiterCsv(recruiterCsvPath));
  if (atsJsonPath) evidence.push(...extractFromAtsJson(atsJsonPath));
  if (notesTxtPath) evidence.push(...extractFromNotesTxt(notesTxtPath));

  const canonical = mergeEvidence(evidence, candidateId);
  const projectionConfig = config || defaultProjectionConfig();
  const projected = project(canonical, projectionConfig);

  let validationErrors: string[] = [];
  if (!config) {
    validationErrors = validateDefaultProfile(projected);
  } else {
    validationErrors = validateCustomProjection(projected, projectionConfig);
  }

  return { canonical, projected, validation_errors: validationErrors };
}

File: src/pipeline.py
from __future__ import annotations

from typing import Any

from .extractors import extract_from_ats_json, extract_from_notes_txt, extract_from_recruiter_csv
from .merger import merge_evidence
from .projector import default_projection_config, project
from .validator import validate_custom_projection, validate_default_profile


def run_pipeline(
    recruiter_csv_path: str | None = None,
    ats_json_path: str | None = None,
    notes_txt_path: str | None = None,
    config: dict[str, Any] | None = None,
    candidate_id: str = "candidate-001",
) -> dict[str, Any]:
    evidence = []
    evidence.extend(extract_from_recruiter_csv(recruiter_csv_path))
    evidence.extend(extract_from_ats_json(ats_json_path))
    evidence.extend(extract_from_notes_txt(notes_txt_path))

    canonical = merge_evidence(evidence, candidate_id=candidate_id)
    projection_config = config if config is not None else default_projection_config()
    projected = project(canonical, projection_config)

    validation_errors: list[str]
    if config is None:
        validation_errors = validate_default_profile(projected)
    else:
        validation_errors = validate_custom_projection(projected, projection_config)

    return {
        "canonical": canonical,
        "projected": projected,
        "validation_errors": validation_errors,
    }
